### PR TITLE
[quant][graphmode][fx] Add graph mode quantization on fx

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43335 [quant][graphmode][fx] Add support for batchnorm relu
* #43334 [quant][graphmode][fx] Add support for batchnorm
* #43333 [quant][graphmode][fx] Add support for cat
* #43332 [quant][graphmode][fx] Add support for add relu
* #43331 [quant][graphmode][fx] Add support for add
* #43287 [quant][graphmode][fx] Add support for conv module + relu
* #43286 [quant] Make OP_LIST_TO_FUSER_METHOD public
* #43285 [quant][graphmode][fx] Add support for conv module
* #43229 [quant][graphmode][fx][test] Add per op test for graph mode quant on fx
* **#43175 [quant][graphmode][fx] Add graph mode quantization on fx**

Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

### CODE REVIEW

**Review by vkuzo — APPROVED** [review]
awesomesauce

we can iterate in future PRs.  One thing which would be great to add is docblocks on public APIs, and a design document for the quantization integration on top of GraphModule

File: test/quantization/test_quantize_fx.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+# eager mode quantization
+from torch.quantization import default_qconfig
+
+# test utils
+from torch.testing._internal.common_quantization import (
+    QuantizationTestCase,
+)
+
+class TestQuantizeFx(QuantizationTestCase):
+    def test_functional(self):
```
Comment (vkuzo) [inline test/quantization/test_quantize_fx.py:22]: would be awesome to also have a basic QAT test case

File: torch/quantization/fx/fuse.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    register_fusion_pattern,
+    get_fusion_patterns,
+)
+
+from .utils import _parent_name
+
+import copy
+
+# Fusion Patterns
+@register_fusion_pattern((torch.nn.BatchNorm2d, torch.nn.Conv2d))
```
Comment (vkuzo) [inline torch/quantization/fx/fuse.py:31]: haven't looked at how this is used, but one question I have is if we can make the order of the pattern follow the order of the layers in the model (i.e. Conv-BN instead of BN-Conv)

File: torch/quantization/fx/quantize.py [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+        self.all_tensors = True
+
+    @abstractmethod
+    def convert(self, quantizer, node, load_arg, debug=False):
+        return NotImplemented
+
+@register_quant_pattern(operator.add)
+@register_quant_pattern((torch.nn.ReLU, operator.add))
+@register_quant_pattern((torch.nn.functional.relu, operator.add))
+class Add(QuantizerPattern):
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:88]: I think for all of these an example (input X, output Y) would be helpful for people to understand the API without having to read every single line of code

File: torch/quantization/fx/fuse.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        for node in reversed(graph.nodes):
+            if node.name not in match_map:
+                for pattern, value in patterns.items():
+                    if matches(modules, node, pattern):
+                        apply_match(pattern, node, (node, value(self, node)))
+
+        return match_map
+
+def fuse(graph_module, inplace=False):
```
Comment (supriyar) [inline torch/quantization/fx/fuse.py:153]: Where do we call the `fuse()` function?

File: test/quantization/test_quantize_fx.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:22]: currently I only have integration tests with torchvision, plan to add it after ddp test works. 
also it will be added in the future when we add individual op tests as well.

File: torch/quantization/fx/fuse.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/fuse.py:31]: that is doable, but the matching is from later ops to former ops, it will confusing when people add new fusion patterns, I feel it would be better to keep the current order considering how matching works right now.

File: torch/quantization/fx/quantize.py [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+
+class DefaultQuant(QuantizerPattern):
+    def convert(self, quantizer, node):
+        assert self.all_tensors
+        return quantize(quantizer, node)
+
+# 2. Post Training Dynamic Quantizatoin Patterns
+@register_dynamic_pattern(torch.nn.Linear)
+@register_dynamic_pattern(torch.nn.functional.linear)
+class DynamicLinear(QuantizerPattern):
```
Comment (supriyar) [inline torch/quantization/fx/quantize.py:370]: For dynamic quant, can we run subset of the graph to run only weight observers?

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:88]: Sure

File: torch/quantization/fx/quantize.py [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+                    bias = kwargs['bias']
+                    kwargs.pop('bias')
+                prepack_args = [weight, bias]
+                packed_weight = quantizer.quantized_graph.call_function(torch.ops.quantized.linear_prepack, prepack_args, {})
+                # construct dynamic linear input
+                linear_input = load_arg(quantized=False)(self.linear_node.args[0])
+                qdynamic_linear_args = [linear_input, packed_weight]
+                return quantizer.quantized_graph.call_function(torch.ops.quantized.linear_dynamic, qdynamic_linear_args, kwargs)
+
+class Quantizer:
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:417]: since these pattern also contain implementations of the feature I feel it might make more sense to put them together. see reference to `quantizer` in the `convert` function of the pattern. the pattern can be considered to be 'tied' to the quantizer implementation.

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:88]: will probably do this in the end after everything is done, since the code might be changing.

File: torch/quantization/fx/quantize.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:377]: This is not handled yet, but should still be doable.

File: torch/quantization/fx/fuse.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/fx/fuse.py:153]: before quantization, there is an example in test/quantization/test_quantize_fx.py

File: torch/quantization/fx/fuse.py (hunk 2 above)
Comment (vkuzo) [reply torch/quantization/fx/fuse.py:31]: cool, then maybe just a comment explaining this would work

File: torch/quantization/fx/quantize.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (supriyar) [reply torch/quantization/fx/quantize.py:428]: Should we consider calling it something other than `Quantizer`? Sine the quantization backend currently uses it for creating different types of qtensors.

File: torch/quantization/fx/fuse.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/fuse.py:31]: had a brief description in the quantize file, I'll move it to pattern_utils.py

File: torch/quantization/fx/quantize.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:426]: yeah, I think so. the official API is not figured out yet, I think maybe we'll not expose this to user and have a quantize_fx as a top level function in the end.

**Comment by jerryzh168** [comment]
cc @jamesr66a @zdevito FYI

File: torch/quantization/fx/quantize.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (raghuramank100) [reply torch/quantization/fx/quantize.py:434]: Second comments here: Request moving the quantizer to a separate file and calling it something else other than Quantizer.

File: torch/quantization/fx/fuse.py (hunk 2 above)
Comment (vkuzo) [reply torch/quantization/fx/fuse.py:27]: wanted to follow up on this, we should discuss defining patterns in the order that they occur in the graph, and changing the matching logic.  Reading through the current implementation, it's not intuitive that the order is reversed.

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (vkuzo) [reply torch/quantization/fx/quantize.py]: I think it's important to do this sooner than later, because the code is harder to read than it needs to be without the docs.  Adding documentation now and adjusting it when things change should be a net benefit, because engineers will spend time reading this code.

File: torch/quantization/fx/fuse.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/fuse.py:27]: sure, I'll change the order

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py]: Sure
